From f9226df1c0a25b0ca0c82607f0b7af47c7ff3f91 Mon Sep 17 00:00:00 2001
From: Stefan Kristiansson <stefan.kristiansson@saunalahti.fi>
Date: Mon, 23 Sep 2013 06:46:15 +0300
Subject: [PATCH 2/2] autoresync on fifo underruns

In case of a fifo underrun, wait for the current frame to end
and then start fetching the new frame.
This will avoid the output from being shifted eternally on
an occasional fifo underrun, instead there will only be a
momentarily 'glitch'.
---
 vga_enh_top.v | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/vga_enh_top.v b/vga_enh_top.v
index 5610624..eacd717 100644
--- a/vga_enh_top.v
+++ b/vga_enh_top.v
@@ -188,6 +188,10 @@ module vga_enh_top (
 	wire [ 8:0] cp_clut_adr;
 	wire [23:0] cp_clut_q;
 
+	// autoresync on fifo underrun;
+	reg wbm_resync;
+	wire wbm_ven;
+
 	//
 	// Module body
 	//
@@ -279,7 +283,7 @@ module vga_enh_top (
 
 		// internal connections
 		.sint        (sint         ),
-		.ctrl_ven    (ctrl_ven     ),
+		.ctrl_ven    (wbm_ven      ),
 		.ctrl_cd     (ctrl_cd      ),
 		.ctrl_vbl    (ctrl_vbl     ),
 		.ctrl_vbsw   (ctrl_vbsw    ),
@@ -402,12 +406,12 @@ module vga_enh_top (
 	);
 
 	// hookup line-fifo
-	wire ctrl_ven_not = ~ctrl_ven;
+	wire wbm_ven_not = ~wbm_ven;
 	vga_fifo_dc #(LINE_FIFO_AWIDTH, 24) line_fifo (
 		.rclk  ( clk_p_i            ),
 		.wclk  ( wb_clk_i           ),
 		.rclr  ( 1'b0               ),
-		.wclr  ( ctrl_ven_not       ),
+		.wclr  ( wbm_ven_not        ),
 		.wreq  ( line_fifo_wreq     ),
 		.d     ( line_fifo_d        ),
 		.rreq  ( line_fifo_rreq     ),
@@ -434,6 +438,14 @@ module vga_enh_top (
 	        luint  <= #1 sluint;      // sample again, reduce metastability risk
 	    end
 
+	// auto resync wbm on fifo underrun
+	assign wbm_ven = ~wbm_resync & ctrl_ven;
+
+	always @(posedge wb_clk_i)
+	  if (vint)
+	    wbm_resync <= 1'b0;
+	  else if (luint & !sluint)
+	    wbm_resync <= 1'b1;
 endmodule
 
 
-- 
1.8.1.2

